include/fmt/compile.h

```diff
@@ -466,10 +466,10 @@ struct is_compiled_format<field<Char, T, N>> : std::true_type {};
 // A replacement field that refers to argument N and has format specifiers.
 template <typename Char, typename T, int N> struct spec_field {
   using char_type = Char;
-  mutable formatter<T, Char> fmt;
+  formatter<T, Char> fmt;
 
   template <typename OutputIt, typename... Args>
-  OutputIt format(OutputIt out, const Args&... args) const {
+  constexpr OutputIt format(OutputIt out, const Args&... args) {
```

Comment:
All these problems started with `mutable`, GCC thinks that `mutable formatter<T, Char> fmt` in `spec_field` structure cannot be used at compile-time. As I can see Clang <7.0 also thinks the same way, this is why the side effect of provided PR is that Clang <7.0 with C++17 support can also use `FMT_COMPILE` with specs.

I failed to get a small code example which raises this error, I tried to create a failing example by recreating all significant parts of the error stack trace, but strangely [it works](https://godbolt.org/z/asa8K6). But I can present you the huge example with this error, this example is created by combining all needed headers of {fmt} together (version from this PR, except `compile.h`) and by making everything needed as `constexpr` in `compile.h` without `mutable`-related changes. Here it is on Compiler Explorer: https://godbolt.org/z/n1Kz38

Since the behaviour of GCC 10.2 and Clang <7.0 is kind of similar, I also tried to recreate the error with Clang 6.0 with my small code example that IMHO should fail - [no luck](https://godbolt.org/z/xvdMaf), so looks like I'm missing something here. But of course, it fails to compile with a bit modified huge example where all {fmt} headers are combined: https://godbolt.org/z/P17TGG

So, this became non-const to eliminate `mutable`, `const CompiledFormat& cf` became non-const references because method `format` is no more `const`, `constexpr auto compiled` became just `auto compiled` because const object cannot be passed via non-const reference. This is my explanation for why I made these changes, of course, I can miss some more elegant solution, especially considering that I failed to recreate this error in a small example. In case it's true and more elegant solution exist I'll be glad to implement it.

Reply:
I think the correct solution is to make `formatter::format` const and keep this `format` function const.

Reply:
Ok, by in my opinion, there is a potential problem with this solution, with making `format()` function in formatters const - maybe if a some of formatters have `format()` function non-const on purpose then it cannot be used at compile-time. I just don't know:

1. why all or almost all formatters currently defined in {fmt} has non-const `format()` function?
2. how rare it is that `format()` function in custom formatters is non-const on purpose?

I mean, I can definitely try to make this function const for formatters which are used in test, but since you have better perspective here it will be helpful if you agree or disagree with me for this potential problem.

Actually, I don't even consider to go this way because some already created formatters would be unavailable at compile-time. On the other hand the code I have in this PR right now does not change formatters at all.

Reply:
> why all or almost all formatters currently defined in {fmt} has non-const format() function?

No good reason, just less typing =).

> how rare it is that format() function in custom formatters is non-const on purpose?

I don't think it should necessary at all. `format` shouldn't mutate `formatter`, only apply format specifiers stored in it to the input.

> already created formatters would be unavailable at compile-time

We can fix this for all {fmt} formatters and later provide a workaround for user-defined ones.

Reply:
Ok, I reverted back most of the changes in `compile.h` and mark all of `formatter`s as const, at least those I found.

> and later provide a workaround for user-defined ones.

But there is no such thing as workarounds at compile-time 🙂, that why it's so cool to check your code just by making it execute at compile-time. Of course, I'm talking about workarounds that are about removing the constness of objects, not interface workarounds.

Reply:
The workaround is to copy the formatter when needed.